com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs

```diff
@@ -219,261 +552,876 @@ internal void OnSceneSwitch(uint sceneIndex, Guid switchSceneGuid, Stream object
+
+        /// <summary>
+        /// Client and Server:
+        /// Generic on scene loaded callback method to be called upon a scene loading
+        /// </summary>
+        private void OnSceneLoaded(string sceneName)
         {
-            CurrentActiveSceneIndex = SceneNameToIndex[s_NextSceneName];
-            var nextScene = SceneManager.GetSceneByName(s_NextSceneName);
-            SceneManager.SetActiveScene(nextScene);
+            var nextScene = GetAndAddNewlyLoadedSceneByName(sceneName);
+            if (!nextScene.isLoaded || !nextScene.IsValid())
+            {
+                throw new Exception($"Failed to find valid scene internal Unity.Netcode for {nameof(GameObject)}s error!");
+            }
+
+            if (SceneEventData.LoadSceneMode == LoadSceneMode.Single)
+            {
+                SceneManager.SetActiveScene(nextScene);
+            }
 
             //Get all NetworkObjects loaded by the scene
-            PopulateScenePlacedObjects();
+            PopulateScenePlacedObjects(nextScene);
 
-            // Move all objects to the new scene
-            MoveObjectsToScene(nextScene);
+            if (SceneEventData.LoadSceneMode == LoadSceneMode.Single)
+            {
+                // Move all objects to the new scene
+                MoveObjectsToScene(nextScene);
+            }
 
+            // The Condition: While a scene is asynchronously loaded in single loading scene mode, if any new NetworkObjects are spawned
+            // they need to be moved into the do not destroy temporary scene
+            // When it is set: Just before starting the asynchronous loading call
+            // When it is unset: After the scene has loaded, the PopulateScenePlacedObjects is called, and all NetworkObjects in the do
+            // not destroy temporary scene are moved into the active scene
             IsSpawnedObjectsPendingInDontDestroyOnLoad = false;
 
-            CurrentSceneIndex = CurrentActiveSceneIndex;
-
             if (m_NetworkManager.IsServer)
             {
-                OnServerLoadedScene(switchSceneGuid);
+                OnServerLoadedScene(nextScene);
             }
             else
             {
-                OnClientLoadedScene(switchSceneGuid, objectStream);
+                // For the client, we make a server scene handle to client scene handle look up table
+                if (!ServerSceneHandleToClientSceneHandle.ContainsKey(SceneEventData.SceneHandle))
+                {
+                    ServerSceneHandleToClientSceneHandle.Add(SceneEventData.SceneHandle, nextScene.handle);
+                }
+                else
+                {
+                    // If the exact same handle exists then there are problems with using handles
+                    throw new Exception($"Server Scene Handle ({SceneEventData.SceneHandle}) already exist!  Happened during scene load of {nextScene.name} with Client Handle ({nextScene.handle})");
```

Comment:
Given this can throw an exception is there any way of moving this check earlier in the process?